### PR TITLE
new lint: `doc_comment_double_space_linebreaks`

### PR DESCRIPTION
**Pull request by Jacherr**
Fixes https://github.com/rust-lang/rust-clippy/issues/12163

I decided to initially make this a restriction lint because it felt a bit niche and opinionated to be a warn-by-default style lint. It may be appropriate as a style lint if the standard or convention *is* to use `\` as doc comment linebreaks - not sure if they are!

The wording on the help message could be improved, as well as the name of the lint itself since it's a bit wordy - suggestions welcome.

This lint works on both `///` and `//!` doc comments.

changelog: new lint: `doc_comment_double_space_linebreaks`

### CODE REVIEW

File: clippy_lints/src/doc/mod.rs [hunk 1]
```diff
@@ -420,6 +421,39 @@ declare_clippy_lint! {
...
+    ///
+    /// ### Why is this bad?
+    /// Double spaces, when used as doc comment linebreaks, can be difficult to see, and may
+    /// accidentally be removed during automatic formatting or manual refactoring. The use of a back-slash (\)
+    /// is clearer in this regard.
+    ///
+    /// ### Example
+    /// The two dots in this example represent a double space.
+    /// ```no_run
+    /// /// This command takes two numbers as inputs and..
```
Comment (Marcono1234) [inline clippy_lints/src/doc/mod.rs:436]: What about using one of the substitute images listed on https://en.wikipedia.org/wiki/Whitespace_character#Substitute_images instead of `.`, e.g. `·` or `␣` (not sure how well this is shown for all fonts though)

File: clippy_lints/src/doc/doc_comment_double_space_linebreak.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                && let comment = sym.as_str()
+                && comment.ends_with("  ")
+            {
+                let pre = match attr.style {
+                    AttrStyle::Outer => "///",
+                    AttrStyle::Inner => "//!",
+                };
+
+                let len = comment.len();
+                let new_comment = format!("{pre}{} \\", comment[..len - 2].to_owned());
```
Comment (Marcono1234) [inline clippy_lints/src/doc/doc_comment_double_space_linebreak.rs:45]: Maybe there should not be a space in front of the `\` because that seems to be preserved, which might be undesirable.
See for example https://spec.commonmark.org/dingus/?text=foo%20%5C%0Abaz%0A, especially the AST tab on the right where this can be seen pretty well.
```suggestion
                let new_comment = format!("{pre}{}\\", comment[..len - 2].to_owned());
```

**Review by Marcono1234 — COMMENTED** [review]
Thanks a lot for implementing the lint I suggested!

I have added two small comments here, but feel free to consider them only suggestions since I am not a project member.

File: clippy_lints/src/doc/doc_comment_double_space_linebreak.rs (hunk 2 above)
Comment (Jacherr) [reply clippy_lints/src/doc/doc_comment_double_space_linebreak.rs:45]: Good point, this is just due to the fact I usually include it :P. I can remove it for sure, and if that's more common it would definitely make more sense to do so.

File: clippy_lints/src/doc/mod.rs (hunk 1 above)
Comment (Jacherr) [reply clippy_lints/src/doc/mod.rs:436]: Thanks for this list, yes, it would make more sense to use a proper substitution character. I think `·` is quite common so I'll go with that.

File: clippy_lints/src/doc/doc_comment_double_space_linebreak.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                && let comment = sym.as_str()
+                && comment.ends_with("  ")
+            {
+                let pre = match attr.style {
+                    AttrStyle::Outer => "///",
+                    AttrStyle::Inner => "//!",
+                };
+
+                let len = comment.len();
+                let new_comment = format!("{pre}{}\\", &comment[..len - 2]);
```
Comment (xFrednet) [inline clippy_lints/src/doc/doc_comment_double_space_linebreak.rs:45]: Since you're indexing into a string here, can you also add a test with multibyte characters? The code should handle the case just fine, but it's always good to test these things. Something like:

```
/// 🌹 are 🟥  
/// 🌷 are 🟦  
/// 📎 is 😎  
/// and so are 🫵
fn multi_byte_chars_tada() {}
```

File: clippy_lints/src/doc/doc_comment_double_space_linebreak.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+fn collect_doc_replacements(attrs: &[Attribute]) -> Vec<(Span, String)> {
+    attrs
+        .iter()
+        .filter_map(|attr| {
+            if let AttrKind::DocComment(com_kind, sym) = attr.kind
+                && com_kind == CommentKind::Line
+                && let comment = sym.as_str()
+                && comment.ends_with("  ")
```
Comment (xFrednet) [inline clippy_lints/src/doc/doc_comment_double_space_linebreak.rs:37]: Somewhere in this lint, we should also check that the item or doc attributes don't come from macro expansion. I think this iterator might be the right place. A simple `!attr.span.from_expansion()` should do the trick.

It would also be cool if you could craft a test example.

**Review by xFrednet — COMMENTED** [review]
Very nice start! I left some comments, mostly for some additional tests. Once they have been addressed, this should be ready to be merged :D

Also thank you for the suggestions from @Marcono1234!

**Comment by Jacherr** [comment]
That's quite a lot of cases to handle - I'm not sure building in a subset of a markdown parser just for this one lint is the right way forward? I suppose it;s only a handful of cases, but seems like it'd be a lot of work to implement and test properly, for the sake of this lint. If push comes to shove I can do it, but perhaps there's an easier/more general way to cover these cases?

**Comment by Jacherr** [comment]
I'm trying to use the events to do this, but even though according to the markdown you provided most of those aren't valid linebreaks, I still seem to be getting HardBreak events for them? It's entirely possible I'm doing something wrong, but it doesn't seem to automatically "distinguish" between a valid linebreak and an invalid one.

**Comment by Jacherr** [comment]
Given the fact that this is pretty niche, and that this is a restriction lint, and the pulldown crate is not up-to-date (with no simple way of updating it), I think the best course of action is leaving this as a FIXME in the lint for future improvement? 

@xFrednet any thoughts on this? Personally I feel like this is a niche enough issue to not be too important, since the double spaces in these cases do nothing and don't need to be there at all in tables etc.

**Comment by xFrednet** [comment]
The changes from the pull request will need some time to get synced back to this PR (Should be ~2 weeks). For now you can edit the `Crates.toml` file here to use the updated version.

@notriddle Thank you for the update and comments! :heart:

**Comment by xFrednet** [comment]
> @xFrednet any thoughts on this? Personally I feel like this is a niche enough issue to not be too important, since the double spaces in these cases do nothing and don't need to be there at all in tables etc.

It depends a bit on how much work it is. Personally, I would prefer it to use `pulldown` since we have it. It will likely also be more maintainable.

**Comment by xFrednet** [comment]
Hey @Jacherr, this is a ping from triage, since there hasn't been any activity in some time. The new version should now be available if you rebase on master.

Do you plan on continueing this implementation?


@rustbot author

**Comment by Jacherr** [comment]
Hey, apologies for the silence, been rather busy the last few weeks. I will complete this implementation but unfortunately will be a week or two before I get to it, if that’s okay.

**Comment by Jacherr** [comment]
I *almost* have this working properly, the one case I'm struggling to handle is detecting if a doc comment is in a macro definition, which we don't want to lint on. Is there any way to detect this from a span?

**Comment by xFrednet** [comment]
@Jacherr  You should be able to call [`span.from_expansion()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.Span.html#method.from_expansion). That's how we usually check for macro expansion.

**Comment by Jacherr** [comment]
@rustbot ready

Hopefully got enough test-cases and caught any edge cases now. I'm not sure if how I extended `check_doc` is considered "correct" - it can pretty easily be changed if needed, I imagine.

File: clippy_lints/src/doc/mod.rs [hunk 5]
```diff
@@ -569,6 +605,7 @@ fn check_attrs(cx: &LateContext<'_>, valid_idents: &FxHashSet<String>, attrs: &[
         return None;
     }
 
+    //doc_comment_double_space_linebreak::check(cx, attrs);
```
[the remark this answers is not in the document]
Comment (Jacherr) [reply clippy_lints/src/doc/mod.rs:608]: Apologies, completely forgot to remove this.

File: clippy_lints/src/doc/doc_comment_double_space_linebreak.rs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+use super::DOC_COMMENT_DOUBLE_SPACE_LINEBREAK;
+
+pub fn check(cx: &LateContext<'_>, attrs: &[Attribute]) {
+    let replacements: Vec<_> = collect_doc_replacements(attrs);
+
+    if let Some((&(lo_span, _), &(hi_span, _))) = replacements.first().zip(replacements.last()) {
+        span_lint_and_then(
+            cx,
+            DOC_COMMENT_DOUBLE_SPACE_LINEBREAK,
+            lo_span.to(hi_span),
```
[the remark this answers is not in the document]
Comment (Jacherr) [reply clippy_lints/src/doc/doc_comment_double_space_linebreak.rs:59]: I imagine this will only work if every line is linted individually. This could be done but not sure it's desired behaviour as it will drastically increase the amount of warnings per paragraph, usually.

File: clippy_lints/src/doc/doc_comment_double_space_linebreak.rs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            "doc comments should use a back-slash (\\) instead of a double space to indicate a linebreak",
+            |diag| {
+                diag.multipart_suggestion(
+                    "replace this double space with a back-slash",
+                    replacements,
+                    Applicability::MachineApplicable,
+                );
+            },
+        );
```
[the remark this answers is not in the document]
Comment (Jacherr) [reply clippy_lints/src/doc/doc_comment_double_space_linebreak.rs:25]: Yes this all makes sense, much better worded than my original one :)

File: clippy_lints/src/doc/mod.rs [hunk 8]
```diff
@@ -422,6 +424,39 @@ declare_clippy_lint! {
...
+    /// /// This command takes two numbers as inputs and··
+    /// /// adds them together, and then returns the result.
+    /// fn add(l: i32, r: i32) -> i32 {
+    ///     l + r
+    /// }
+    /// ```
+    ///
+    /// Use instead:
+    /// ```no_run
+    /// /// This command takes two numbers as inputs and \
```
Comment (Marcono1234) [inline clippy_lints/src/doc/mod.rs:448]: Should probably not have a space, see also https://github.com/rust-lang/rust-clippy/pull/12876#discussion_r1623250888
```suggestion
    /// /// This command takes two numbers as inputs and\
```

File: clippy_lints/src/doc/mod.rs [hunk 9]
```diff
@@ -422,6 +424,39 @@ declare_clippy_lint! {
     "require every line of a paragraph to be indented and marked"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects doc comment linebreaks that use double spaces to separate lines, instead of back-slash (\).
+    ///
+    /// ### Why is this bad?
+    /// Double spaces, when used as doc comment linebreaks, can be difficult to see, and may
+    /// accidentally be removed during automatic formatting or manual refactoring. The use of a back-slash (\)
```
Comment (Marcono1234) [inline clippy_lints/src/doc/mod.rs:433]: Assuming this is a standard Rust doc comment[^1], the `\` probably has to be escaped, otherwise it currently (redundantly) escapes the `)`, and is rendered as "()":
```suggestion
    /// Detects doc comment linebreaks that use double spaces to separate lines, instead of back-slash (\\).
    ///
    /// ### Why is this bad?
    /// Double spaces, when used as doc comment linebreaks, can be difficult to see, and may
    /// accidentally be removed during automatic formatting or manual refactoring. The use of a back-slash (\\)
```

Or alternatively could format it as code:
```suggestion
    /// Detects doc comment linebreaks that use double spaces to separate lines, instead of back-slash (`\`).
    ///
    /// ### Why is this bad?
    /// Double spaces, when used as doc comment linebreaks, can be difficult to see, and may
    /// accidentally be removed during automatic formatting or manual refactoring. The use of a back-slash (`\`)
```

[^1]: Not sure if the clippy documentation generation treats this in some special way.

File: clippy_lints/src/doc/doc_comment_double_space_linebreak.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (xFrednet) [reply clippy_lints/src/doc/doc_comment_double_space_linebreak.rs]: I would be in favor of having a warning for each double space. Looking at the [lintcheck output](https://github.com/rust-lang/rust-clippy/actions/runs/10331222357?pr=12876) in our CI, it's hard to see what the actual problem is.

In this window it looks like a normal line break:

```
   --> target/lintcheck/sources/arrayvec-0.7.4/src/arrayvec.rs:380:53
    |
380 |       /// This is a checked version of `.swap_remove`.  
    |  _____________________________________________________^
381 | |     /// This operation is O(1).
    | |________^
```

If the message only spanned the two spaces, it would look like this:

```
   --> target/lintcheck/sources/arrayvec-0.7.4/src/arrayvec.rs:380:53
    |
380 |       /// This is a checked version of `.swap_remove`.  
    |                                                       ^^
    |
```

This is much clearer IMO.

It also makes it easier to apply the suggestion automatically as there is less chance to have the span overlap with other lint emissions.

File: clippy_lints/src/doc/mod.rs (hunk 9 above)
Comment (xFrednet) [reply clippy_lints/src/doc/mod.rs:433]: It is a standard Rust doc comment. The output is later rendered with a JS Markdown thingy, but it's likely that a double backslash is needed. Using backticks to make it inline code works and is the way I would recommend.

(@Marcono1234, thank you for the comment, I would have probably missed this :D)

**Review by xFrednet — COMMENTED** [review]
Two smaller comments, but this implementation already looks good :D

Also, big thanks to @notriddle and @Marcono1234 for their reviews :heart:

File: clippy_lints/src/doc/mod.rs (hunk 8 above)
Comment (Marcono1234) [reply clippy_lints/src/doc/mod.rs:557]: @Jacherr, I think this is still open, or did you intentionally keep it this way?

Comment (Jacherr) [reply clippy_lints/src/doc/mod.rs]: Yep, somehow missed this 😅. Should be resolved now.

**Comment by Jacherr** [comment]
@xFrednet this has been sat for a while, does all look good on your end? Seems that all that was left were a few nits which I cleaned up so once the conflicts are solved would this be ready to be merged? No rush of course

**Comment by xFrednet** [comment]
Woops, this totally slipped by my inbox. Sorry, thank you for pinging me! :sweat_smile: 

I'll put it in my reviewing queue. Currently it takes a bit longer due to RL stuff, but you should have a review in 1-2 weeks.

File: tests/ui/doc/doc_comment_double_space_linebreak.stderr [hunk 10]
```diff
@@ -0,0 +1,76 @@
+error: doc comment uses two spaces for a hard line break
+  --> tests/ui/doc/doc_comment_double_space_linebreak.rs:7:43
+   |
+LL |   //! Should warn on double space linebreaks  
+   |  ___________________________________________^
+LL | | //! in file/module doc comment
+   | |____^
+   |
```
Comment (xFrednet) [inline tests/ui/doc/doc_comment_double_space_linebreak.stderr:8]: Have you tried changing the error span, like suggested here?

https://github.com/rust-lang/rust-clippy/pull/12876#discussion_r1713615334

```suggestion
  --> tests/ui/doc/doc_comment_double_space_linebreak.rs:7:43
   |
LL |   //! Should warn on double space linebreaks  
   |                                             ^^
   |
   |
```

It should be something simple, like taking a span from `lo_span` to `lo_span + 2`

**Review by xFrednet — COMMENTED** [review]
I believe a previous comment is still unresolved, at least according to the `.stderr` file, otherwise LGTM

**Comment by Jacherr** [comment]
Just to be clear - is this fine as a restriction lint? I have no idea if the preferred way is to use a backslash or double space. If the correct method is a backslash maybe this would be better as a warn by default?

**Comment by xFrednet** [comment]
> Just to be clear - is this fine as a restriction lint?

Yes, I think this category is perfect for the lint. That's why I didn't comment on it further.

> I have no idea if the preferred way is to use a backslash or double space. If the correct method is a backslash maybe this would be better as a warn by default?

I think a backslash should be the default if you need a random line break, for the reasons described in the lint description.

**Comment by Jacherr** [comment]
So just to be completely sure, it's in pedantic right now and that's all good? Just don't want to miss anything before getting it looked at again and hopefully merged 😅

**Comment by xFrednet** [comment]
Currently, it's a pedantic lint. I'll ask for feedback on the category as part of the Final Comment Period (FCP) :)

**Comment by xFrednet** [comment]
Hej @Jacherr, I wanted to play a bit around with the error message to have the best of both worlds, one lint per paragraph, but also just the two spaces highlighted. I pushed my fix now and also adjusted the lint name, by adding an s to the end.

Now this should be good to go, I'll create an FCP and also ask for feedback on the lint category.

(I hope it was okay to push to your branch and also squash your commits :see_no_evil: )

**Comment by xFrednet** [comment]
Roses are red,
FCP is done,
pedantic group was selected,
let the CI run!
